sound/soc/sof/ipc.c

```diff
@@ -654,6 +664,133 @@ static void sof_set_get_mmap(struct snd_sof_dev *sdev,
+		sipc->hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		sipc->elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		sipc->msg_bytes = scontrol->num_channels *
+			sizeof(struct sof_ipc_ctrl_value_comp);
+		sipc->hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		sipc->elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_DATA_GET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sipc->msg_bytes = cdata->data->size;
+		sipc->hdr_bytes = sizeof(struct sof_ipc_ctrl_data) +
+			sizeof(struct sof_abi_hdr);
+		sipc->elems = cdata->data->size;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	cdata->rhdr.hdr.size = sipc->msg_bytes + sipc->hdr_bytes;
+	cdata->num_elems = sipc->elems;
+	cdata->total_elems = sipc->elems;
+
+	return 0;
+}
+
+static int sof_get_ctrl_copy_params(enum sof_ipc_ctrl_type ctrl_type,
+				    struct sof_ipc_ctrl_data *src,
+				    struct sof_ipc_ctrl_data *dst,
+				    struct sof_ctrl_data_send_params *sipc)
+{
+	switch (ctrl_type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		sipc->src = (unsigned char *)src->chanv;
+		sipc->dst = (unsigned char *)dst->chanv;
+		break;
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		sipc->src = (unsigned char *)src->compv;
+		sipc->dst = (unsigned char *)dst->compv;
+		break;
+	case SOF_CTRL_TYPE_DATA_GET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sipc->src = (unsigned char *)src->data->data;
+		sipc->dst = (unsigned char *)dst->data->data;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* calculate payload size and number of messages */
+	sipc->pl_size = SOF_IPC_MSG_MAX_SIZE - sipc->hdr_bytes;
+	sipc->num_msg = DIV_ROUND_UP(sipc->msg_bytes, sipc->pl_size);
+
+	return 0;
+}
+
+static int sof_set_get_large_ctrl_data(struct snd_sof_dev *sdev,
+				       struct sof_ipc_ctrl_data *cdata,
+				       struct sof_ctrl_data_send_params *sipc,
+				       int send)
+{
+	struct sof_ipc_ctrl_data *partdata;
+	u32 send_bytes;
+	u32 offset;
+	int err = 0;
+	int i;
+
+	/* allocate max ipc size because we have at least one */
+	partdata = (struct sof_ipc_ctrl_data *)
+		kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!partdata)
+		return -ENOMEM;
+
+	if (send)
+		sof_get_ctrl_copy_params(cdata->type, cdata, partdata, sipc);
+	else
+		sof_get_ctrl_copy_params(cdata->type, partdata, cdata, sipc);
+
+	/* copy the header data */
+	memcpy(partdata, cdata, sipc->hdr_bytes);
+
+	/* copy the data in a loop */
+	offset = 0;
+	for (i = 0; i < sipc->num_msg; i++) {
+		partdata->msg_id = i;
+		send_bytes = sipc->msg_bytes > sipc->pl_size ? sipc->pl_size :
+			sipc->msg_bytes;
+		partdata->num_elems = send_bytes;
+		partdata->rhdr.hdr.size = sipc->hdr_bytes + send_bytes;
+		if (send)
+			memcpy(sipc->dst, sipc->src + offset, send_bytes);
+		err = sof_ipc_tx_message(sdev->ipc,
+					 partdata->rhdr.hdr.cmd,
+					 partdata,
+					 partdata->rhdr.hdr.size,
+					 partdata,
+					 partdata->rhdr.hdr.size);
```

Comment:
test on err? You really want to stop sending if there is one error. Continuing and checking the last IPC could conceal an intermediate error.

Reply:
oh, this was lost in rebase. fixed now.